iib/web/status.py

```diff
@@ -0,0 +1,140 @@
+            return False, "an error occurred"
+
+    return True, None
+
+
+def _ping_celery_workers(retries):
+    """
+    Attempt to ping worker, retry on ConnectionError.
+
+    :param int retries: number of retry/retries before returning an empty response
+    :return: dict of {hostname: reply} as returned by ping()
+    """
+    for i in range(retries + 1):
+        if i > 0:
+            time.sleep(0.25 * 2 ** (i - 1))
+        try:
+            replies = app.control.inspect().ping()
+        except ConnectionError:
+            continue
+
+        return replies or {}
+
+    return {}
+
+
+def workers_status(retries=2):
+    """
+    Ping workers, check received replies.
+
+    :param int retries: number of retry/retries on ConnectionError when pinging workers
+    :return: list of workers status information (empty if no replies)
+    """
+    replies = _ping_celery_workers(retries=retries)
+
+    # replies is a dict of {hostname: reply}, convert to a sorted list of [(hostname, reply)]
+    reply_tuples = sorted(replies.items(), key=lambda kv: kv[0])
+    workers = []
+
+    for worker_name, reply in reply_tuples:
+        # the reply is in the following format {'ok': 'pong'}.
+        worker = {"name": worker_name, "ok": "ok" in reply}
+
+        if "ok" not in reply:
+            worker["reason"] = reply.get("error", "unknown reason")
+
+        workers.append(worker)
+
+    return workers
+
+
+def status(short=False, worker_ping_retries=2):
+    """
+    Get status of IIB workers and services that IIB depends on.
+
+    Display the current version of podman, skopeo, opm.
+    :param bool short: raise an error as soon as any problem is found
+    :param int worker_ping_retries: number of retry/retries on ConnectionError when pinging workers
+    :return: dict with the following keys:
+        "services": list of status info for individual services
+        "software": list of softwares and their versions
+        "workers": list of status info for individual workers
+    :raises IIBError: if short is True and a problem is found
+    """
+    services = []
+
+    def add_status(service_name, ok, reason):
+        if short and not ok:
+            raise IIBError(f"{service_name} unavailable: {reason}")
+        service = {"name": service_name, "ok": ok}
+        if not ok:
+            service["reason"] = reason
+        services.append(service)
+
+    add_status("DATABASE", *database_connection_check())
+    rabbit_ok, rabbit_reason = rabbitmq_connection_check(get_worker_config().broker_url)
+    add_status("RABBITMQ", rabbit_ok, rabbit_reason)
+
+    if rabbit_ok:
+        workers = workers_status(retries=worker_ping_retries)
+    else:
+        workers = []
+
+    any_worker_ok = any(worker["ok"] for worker in workers)
+    if short and not any_worker_ok:
+        raise IIBError("no workers are available")
+
+    return {
+        "status": services,
```

Comment:
I am not sure if we can change type of "status". Until now it is plain text, however with this change we are returning dictionary. It can break parsing of the response.